thrust/thrust/zip_function.h

```diff
@@ -179,6 +179,36 @@ class zip_function
   mutable Function func;
 };
 
+//! @brief The invoke machinery we are using for the generic case does not work with `proclaim_return_type` because
+//! that obviously tries to deduce the return type from the device lambda, which we need `proclaim_return_type` for
+template <class _Ret, class _DecayFn>
+class zip_function<::cuda::__detail::__return_type_wrapper<_Ret, _DecayFn>>
+{
+  using Function = ::cuda::__detail::__return_type_wrapper<_Ret, _DecayFn>;
+
+public:
+  zip_function() = default;
+
+  _CCCL_HOST_DEVICE zip_function(Function func)
+      : func(_CUDA_VSTD::move(func))
+  {}
+
+  template <typename Tuple>
+  _CCCL_HOST_DEVICE _Ret operator()(Tuple&& args) const
+  {
+    return func(THRUST_FWD(args));
+  }
```

Comment:
Question: I don't understand how this can work. The purpose of `zip_function` is to turn a `tuple<Ts...> args` into a call to `func(args[Is]...)`. We are missing the `apply` here somewhere, don't we? Or is this code never executed anyway and only required for type deduction?

Reply:
I definitely should not work late in the evening. Obviously this needs to call apply